src/VisualStudio/Core/Def/TaskList/ExternalErrorDiagnosticUpdateSource.cs

```diff
@@ -139,385 +110,161 @@ public ImmutableArray<DiagnosticData> GetBuildErrors()
     public bool IsSupportedDiagnosticId(ProjectId projectId, string id)
         => GetBuildInProgressState()?.IsSupportedDiagnosticId(projectId, id) ?? false;
 
-    private void OnBuildProgressChanged(InProgressState? state, BuildProgress buildProgress)
+    public void ClearErrors(ProjectId projectId)
     {
-        if (state != null)
+        // Clear the previous errors associated with the project.
+        _taskQueue.AddWork(async cancellationToken =>
         {
-            _lastBuiltResult = state.GetBuildErrors();
-        }
-
-        RaiseBuildProgressChanged(buildProgress);
+            await ClearPreviousAsync(projectId: projectId, cancellationToken).ConfigureAwait(false);
+        });
     }
 
-    public void ClearErrors(ProjectId projectId)
+    internal void OnSolutionBuildStarted()
     {
-        // Capture state if it exists
-        var state = GetBuildInProgressState();
-
-        // Update the state to clear diagnostics and raise corresponding diagnostic updated events
-        // on a serialized task queue.
-        _taskQueue.ScheduleTask(nameof(ClearErrors), async () =>
-        {
-            if (state == null)
-            {
-                // TODO: Is it possible that ClearErrors can be invoked while the build is not in progress?
-                // We fallback to current solution in the workspace and clear errors for the project.
-                await ClearErrorsCoreAsync(projectId, _workspace.CurrentSolution, state).ConfigureAwait(false);
-            }
-            else
-            {
-                // We are going to clear the diagnostics for the current project.
-                // Additionally, we clear errors for all projects that transitively depend on this project.
-                // Otherwise, fixing errors in core projects in dependency chain will leave back stale diagnostics in dependent projects.
-
-                // First check if we already cleared the diagnostics for this project when processing a referenced project.
-                // If so, we don't need to clear diagnostics for it again.
-                if (state.WereProjectErrorsCleared(projectId))
-                {
-                    return;
-                }
-
-                var solution = state.Solution;
-
-                await ClearErrorsCoreAsync(projectId, solution, state).ConfigureAwait(false);
-
-                var transitiveProjectIds = solution.GetProjectDependencyGraph().GetProjectsThatTransitivelyDependOnThisProject(projectId);
-                foreach (var projectId in transitiveProjectIds)
-                {
-                    if (state.WereProjectErrorsCleared(projectId))
-                    {
-                        continue;
-                    }
-
-                    await ClearErrorsCoreAsync(projectId, solution, state).ConfigureAwait(false);
-                }
-            }
-        }, GetApplicableCancellationToken(state));
-
-        return;
+        _ = GetOrCreateInProgressState();
```

Comment:
this feels a little dangerous without threading guarantees.

Reply:
this is the same as it was before, calling `GetOrCreateInProgressState();` in `OnSolutionBuildStarted`.  It also shouldn't be called concurrently, as `OnSolutionBuildStarted` is called in response to the sln build UI context (can only be 1 running at a time).  Adding a comment.

Reply:
k.  i can live with this.